[Beta]: docs(cn): translate learn/queueing-a-series-of-state-updates (#677)
reactjs/zh-hans.reactjs.org: beta/src/pages/learn/queueing-a-series-of-state-updates.md

---
title: 状态更新的队列
translators: 
  - Jiacheng787
---

<Intro>

> 设置组件状态将会把一个渲染任务加入队列。但有时你可能会希望在下次渲染任务加入队列之前，执行多次状态更新。为此，这有助于了解 React 如何批量更新状态。

Author: 感觉 " React 如何对 state 更新进行批处理" 这句还是有问题，改成 "React 如何批量更新 state" 是否更好。

> 设置组件状态将会把一个渲染任务加入队列。但有时你可能会希望在下次渲染任务加入队列之前，执行多次状态更新。为此，这有助于了解 React 如何批量更新状态。

Reviewer: > 感觉 " React 如何对 state 更新进行批处理" 这句还是有问题，改成 "React 如何批量更新 state" 是否更好。

@Jiacheng787 通顺性上没问题，但可能语义上有一些出入。因为如果要翻译成“批量更新 state”的话，原文就应该是 “batch-update the state”，而不是 “batch state updates”。

> 设置组件状态将会把一个渲染任务加入队列。但有时你可能会希望在下次渲染任务加入队列之前，执行多次状态更新。为此，这有助于了解 React 如何批量更新状态。

Author: 语义上是没问题，但是翻译成中文还是要符合中文的表达，否则读者肯定会觉得很奇怪

> 设置组件状态将会把一个渲染任务加入队列。但有时你可能会希望在下次渲染任务加入队列之前，执行多次状态更新。为此，这有助于了解 React 如何批量更新状态。

Reviewer: @Jiacheng787 你说的也有道理，那我把我的建议改成“批量更新 state”，你就可以直接用了。

> 设置组件状态将会把一个渲染任务加入队列。但有时你可能会希望在下次渲染任务加入队列之前，执行多次状态更新。为此，这有助于了解 React 如何批量更新状态。

Reviewer: @Jiacheng787 改好了，麻烦应用一下，谢谢！

</Sandpack>

> 但是，你可能还记得上一节中的内容，[每一次渲染的状态值都是固定的](/learn/state-as-a-snapshot#rendering-takes-a-snapshot-in-time)，因此无论你调用多少次 `setNumber(1)`，在第一次渲染的事件处理函数内部的 `number` 值总是 `0` ：

Author: > 因此无论你调用多少次 `setNumber(1)`

感觉改成 “因此无论你调用多少次 `setNumber(number + 1)`” 比较合理

</Sandpack>

在这里，`n => n + 1` 被称为 **更新函数**。当你将它传递给 state setter 时：

1. React 将此函数加入队列，在事件处理函数中的所有其他代码运行后进行处理。
2. 在下一次渲染期间，React 遍历队列并提供最终更新状态。

```js
setNumber(n => n + 1);
setNumber(n => n + 1);
setNumber(n => n + 1);
```

下面是 React 在执行事件处理函数时的处理流程：

1. `setNumber(n => n + 1)`：`n => n + 1` 是一个函数。React 将它加入队列。
2. `setNumber(n => n + 1)`：`n => n + 1` 是一个函数。React 将它加入队列。
3. `setNumber(n => n + 1)`：`n => n + 1` 是一个函数。React 将它加入队列。

当你在下次渲染期间调用 `useState`，React 会遍历队列。初始的 `number` 状态是 `0`，所以这就是 React 作为 `n` 参数传递给第一个更新函数的值。然后 React 获取上一个更新函数的返回值，并将其作为 `n` 传递给下一个更新函数，以此类推：

|  更新队列 | `n` | 返回值 |
|--------------|---------|-----|
| `n => n + 1` | `0` | `0 + 1 = 1` |
| `n => n + 1` | `1` | `1 + 1 = 2` |
| `n => n + 1` | `2` | `2 + 1 = 3` |

React 保存 `3` 为最终结果并从 `useState` 中返回。

这就是为什么在上面的示例中点击“+3”正确地将值增加“+3”。

> ### 如果在替换状态后更新状态会发生什么 {/*what-happens-if-you-update-state-after-replacing-it*/}

Author: 这里原文的换行有问题

> ### 如果在替换状态后更新状态会发生什么 {/*what-happens-if-you-update-state-after-replacing-it*/}

Reviewer: ```suggestion
### 如果你在替换 state 后更新 state 会发生什么 {/*what-happens-if-you-update-state-after-replacing-it*/}
```

</Sandpack>

这是事件处理函数告诉 React 要做的事情：

1. `setNumber(number + 5)`：`number` 为 `0`，所以 `setNumber(0 + 5)`。React 将 *“替换为 `5`”* 添加到其队列中。
2. `setNumber(n => n + 1)`：`n => n + 1` 是一个更新函数。 React 将 *该函数* 添加到其队列中。

> 在下一次渲染期间，React 会遍历状态队列：

Reviewer: ```suggestion
在下一次渲染期间，React 会遍历 state 队列：
```

</Sandpack>

以下是 React 在执行事件处理函数时的处理流程：

1. `setNumber(number + 5)`：`number` 为 `0`，所以 `setNumber(0 + 5)`。React 将 *“替换为 `5`”* 添加到其队列中。
2. `setNumber(n => n + 1)`：`n => n + 1` 是一个更新函数。React 将该函数添加到其队列中。
3. `setNumber(42)`：React 将 *“替换为 `42`”* 添加到其队列中。

> 在下一次渲染期间，React 会遍历状态队列：

Reviewer: ```suggestion
在下一次渲染期间，React 会遍历 state 队列：
```

|   更新队列       | `n` | 返回值 |
|--------------|---------|-----|
| “替换为 `5`” | `0`（未使用） | `5` |
| `n => n + 1` | `5` | `5 + 1 = 6` |
| “替换为 `42`” | `6`（未使用） | `42` |

然后 React 保存 `42` 为最终结果并从 `useState` 中返回。

总而言之，以下是你可以考虑传递给 `setNumber` state setter 的内容：

* **一个更新函数**（例如：`n => n + 1`）被添加到队列中。
* **任何其他值**（例如：数字 `5`）将“替换为 `5`”添加到队列中，忽略已经在队列中的内容。

事件处理函数执行完成后，React 将触发重新渲染。在重新渲染期间，React 将处理队列。更新函数在渲染期间执行，因此 **更新函数必须是 [纯函数](/learn/keeping-components-pure)** 并且只 *返回* 结果。不要尝试从它们内部设置状态或者执行其他副作用。在严格模式下，React 会执行每个更新函数两次（但是丢弃第二个结果）以帮助你发现错误。

### 命名约定 {/*naming-conventions*/}

通常通过相应状态变量的第一个字母来命名更新函数参数：

```js
setEnabled(e => !e);
setLastName(ln => ln.reverse());
setFriendCount(fc => fc * 2);
```

如果你喜欢更冗长的代码，另一个常见的约定是重复完整的状态变量名称，如 `setEnabled(enabled => !enabled)`，或使用前缀如 `setEnabled(prevEnabled => !prevEnabled)`。

<Recap>

* 设置状态不会更改现有渲染中的变量，但会请求新的渲染。
* React 在事件处理函数执行完成之后处理状态更新。这称为批处理。

> * 要在一个事件中多次更新某些状态，你可以使用 `setNumber(n => n + 1)` 更新函数。

Reviewer: ```suggestion
* 要在一个事件中多次更新某些 state，你可以使用 `setNumber(n => n + 1)` 更新函数。
```

</Sandpack>

这可以确保当你增加或减少计数器时，会根据其 *最新* 状态而不是点击时的状态来执行此操作。

</Solution>

### 自己实现状态队列 {/*implement-the-state-queue-yourself*/}

在这个挑战中，你将从头开始重新实现 React 的一小部分！这并不像听起来那么难。

滚动 sandbox 进行预览。请注意，它显示了 **四个测试用例** 。它们对应于你之前在本页上看到的示例。你的任务是实现 `getFinalState` 函数，以便它为每种情况返回正确的结果。如果你正确实现它，则所有四个测试用例都应该通过。

你将收到两个参数：`baseState` 是初始状态（例如：`0`），`queue` 是包含数字（例如：`5`）和更新函数（例如：`n => n + 1`）的混合数组，按添加顺序排列。

你的任务是返回最终状态，就像这个页面上的表格一样！

<Hint>

如果你没有思路，可以从下面的代码结构开始:

```js
export function getFinalState(baseState, queue) {
  let finalState = baseState;

  for (let update of queue) {
    if (typeof update === 'function') {
      // TODO: 调用更新函数
    } else {

>       // TODO: 替换状态

Reviewer: ```suggestion
      // TODO: 替换 state
```

    }
  }

  return finalState;
}
```

> 完成剩余的代码！

Reviewer: ```suggestion
补全缺失的几行代码！
```

  const actual = getFinalState(baseState, queue);
  return (
    <>

>       <p>初始状态：<b>{baseState}</b></p>

Reviewer: ```suggestion
      <p>初始 state：<b>{baseState}</b></p>
```

  const actual = getFinalState(baseState, queue);
  return (
    <>

>       <p>初始状态：<b>{baseState}</b></p>

Reviewer: ```suggestion
      <p>初始 state：<b>{baseState}</b></p>
```
